File: io/github/zhengyhn/leetcode/valid_anagram/valid-anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
  if (s.length != t.length) {
    return false;
  }
  var first = Array.apply(null, Array(26)).map(function () { return 0; });
  var second = first.slice();

  for (var i = 0; i < s.length; ++i) {
    first[s[i].charCodeAt(0) - 'a'.charCodeAt(0)] += 1;
  }
  for (var i = 0; i < t.length; ++i) {
    second[t[i].charCodeAt(0) - 'a'.charCodeAt(0)] += 1;
  }

  for (var i = 0; i < 26; ++i) {
    if (first[i] != second[i]) {
      return false;
    }
  }
  return true;
};

console.log(isAnagram('rat', 'cat'));

File: io/github/zhengyhn/leetcode/valid_anagram/valid-anagram.py
class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) != len(t):
            return False

        first = [0] * 26
        second = first[:]
        for c in s:
            first[ord(c) - ord('a')] += 1
        for c in t:
            second[ord(c) - ord('a')] += 1

        for i in range(0, 26):
            if first[i] != second[i]:
                return False

        return True

sln = Solution()
if sln.isAnagram('rat', 'cat'):
    print 't'
else:
    print 'f'
